x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimeoutChecker.java

```diff
@@ -143,6 +141,9 @@ public void register(Matcher matcher) {
                 boolean wasFalse = value.registered.compareAndSet(false, true);
                 assert wasFalse;
                 value.matchers.add(matcher);
+                if (value.isTimedOut()) {
```

Comment:
Doing this check unsynchronized means there's a chance for the matcher to be interrupted twice.

I know you have worked around this in the test by using `atLeastOnce()`, but it leaves the question of what a real Joni `Matcher` does if interrupted twice.

So I think that either you need to investigate what happens if a real Joni `Matcher` is interrupted twice, or (probably simpler) put this bit in a `synchronized` block so that it cannot run while `interruptLongRunningThreadIfRegistered` is part way through.